### PR TITLE
Add success and failure method for Sum type

### PR DESCRIPTION
**Pull request by GustavoCaso**
#184 

After checking the code looks like the `Sum` type was missing the `success` method, creating some errors using `dry-struct`.

This PR adds the `success` method.

### CODE REVIEW

**Comment by solnic** [comment]
Could you also add `failure` so that it can close the related issue too?

**Comment by GustavoCaso** [comment]
@solnic Added!

**Comment by GustavoCaso** [comment]
@solnic any update on this one?

**Comment by solnic** [comment]
Almost there. We need a spec that ensures that the original issue is fixed too. Just make sure something like this works:

``` ruby
Types::Coercible::Int.optional.safe[nil] # should return a success
Types::Coercible::Int.optional.safe['1'] # should return a success
Types::Coercible::Int.optional.safe['foo'] # should return a failure
```

File: spec/dry/types/sum_spec.rb [hunk 1]
```diff
@@ -120,6 +120,28 @@
...
+      expect{
+        type.success('true')
+      }.to raise_error(ArgumentError, /Invalid success value 'true'/)
+    end
+  end
+
+  describe '#failure' do
+    subject(:type) { Dry::Types['int'] | Dry::Types['string'] }
+
+    it 'returns success when value passed' do
```
Comment (solnic) [inline spec/dry/types/sum_spec.rb]: This should be `it 'returns failure when invalid value is passed'`

**Comment by GustavoCaso** [comment]
@solnic I added the correct spec description 😄

**Pull request merged**